src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+        # return the point.
+        if self.containsPoint(point):
+            return point
+
+        # Get first point hit in both directions of ray
+        point = point.coordinates
+
+        if on_direction is not None:
+            on_direction = numpy.array(on_direction)
+        else:
+            if isinstance(self, MeshVolumeRegion):
+                on_direction = numpy.array((0,0,1))
+            elif isinstance(self, MeshSurfaceRegion):
+                on_direction = sum(self.mesh.face_normals*self.mesh.area_faces[:, numpy.newaxis])/self.mesh.area
+                on_direction = on_direction/numpy.linalg.norm(on_direction)
+
+        intersection_data, _, _ = self.mesh.ray.intersects_location(ray_origins=[point, point], \
+            ray_directions=[on_direction, -1*on_direction], multiple_hits=False)
+
+        if len(intersection_data) == 0:
+            return None
+
+        # Get point with least euclidean distance
+        def euclidean_distance(p_1, p_2):
+            diff_list = [p_1[i] - p_2[i] for i in range(3)]
+            square_list = [math.pow(p, 2) for p in diff_list]
+            return math.sqrt(sum(square_list))
+
+        distances = [euclidean_distance(point, p) for p in intersection_data]
+
+        closest_point = intersection_data[distances.index(min(distances))]
+
+        return Vector(*closest_point)
+
+    @cached_property
+    def circumcircle(self):
+        """ Compute an upper bound on the radius of the region"""
+        assert not needsSampling(self)
+
+        center_point = Vector(*self.mesh.bounding_box.center_mass)
+        half_extents = [val/2 for val in self.mesh.extents]
+        circumradius = hypot(*half_extents)
+
+        return (center_point, circumradius)
+
+    @cached_property
+    def isConvex(self):
+        return self.mesh.is_convex
+
+    @property
+    def AABB(self):
+        return (tuple(self.mesh.bounds[0]),tuple(self.mesh.bounds[1]),tuple(self.mesh.bounds[2]))
+
+    @cached_property
+    def _boundingPolygon(self):
+        assert not needsSampling(self)
+
+        # Relatively fast case for convex regions
+        if self.isConvex:
+            return shapely.geometry.MultiPoint(self.mesh.vertices).convex_hull
+
+        # Generic case for arbitrary shapes
+        if self.mesh.is_watertight:
+            projection = trimesh.path.polygons.projected(self.mesh, normal=(0,0,1), rpad=1e-4)
+        else:
+            # Special parameters to use all faces if mesh is not watertight.
+            projection = trimesh.path.polygons.projected(self.mesh, normal=(0,0,1), rpad=1e-4,
+                                                         ignore_sign=False, tol_dot=-float('inf'))
+
+        return projection
+
+    @cached_property
+    def boundingPolygon(self):
+        """A PolygonalRegion bounding the mesh"""
+        return PolygonalRegion(polygon=self._boundingPolygon)
+
+    def sampleGiven(self, value):
```

Comment:
This class (and all the others which allow random arguments) will need a similar `evaluateInner` method.